File: poc-csharp/RpsPoc/Concurrency/Orchestrator.cs
using System.Collections.Concurrent;

using RpsPoc.Domain;

namespace RpsPoc.Concurrency;

/// <summary>
/// Loop proprio: tira duplas da fila e despacha. Nao joga nada.
/// <para>
/// O limite de T matches simultaneos vem do <see cref="SemaphoreSlim"/>, nao de
/// um pool de tamanho fixo: em .NET o pool e global, entao a forma de honrar o
/// RF02 e limitar quantas tasks concorrentes existem.
/// </para>
/// </summary>
public sealed class Orchestrator(
    PlayerQueue queue,
    EventBus bus,
    ConcurrentDictionary<int, Match> activeMatches,
    SemaphoreSlim slots,
    int roundDelayMs)
{
    private int _nextMatchId;

    public async Task RunAsync(CancellationToken cancellationToken)
    {
        try
        {
            while (!cancellationToken.IsCancellationRequested)
            {
                await slots.WaitAsync(cancellationToken).ConfigureAwait(false);

                // DequeuePair bloqueia (Monitor.Wait). Sai pro pool pra nao
                // prender a thread que carrega o orquestrador.
                var pair = await Task.Run(queue.DequeuePair, cancellationToken).ConfigureAwait(false);
                if (pair is null)
                {
                    slots.Release();
                    break; // RF09: sobrou 1, nenhum match ativo
                }

                var match = new Match(Interlocked.Increment(ref _nextMatchId), pair.Value.First, pair.Value.Second);
                activeMatches[match.Id] = match;
                bus.Publish(new QueueChanged(queue.Snapshot(), queue.Alive));

                _ = new MatchRunner(match, queue, bus, activeMatches, slots, roundDelayMs)
                    .RunAsync(cancellationToken);
            }
        }
        catch (OperationCanceledException)
        {
            // encerramento manual
        }
        finally
        {
            bus.Publish(new TournamentEnded(queue.Champion()));
        }
    }
}

File: poc-java/src/main/java/rps/concurrency/Orchestrator.java
package rps.concurrency;

import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

import rps.domain.Match;
import rps.domain.Player;

/**
 * Thread propria, em loop:
 * tira duplas da fila e entrega pro pool.
 * Nao joga nada — so parea e despacha.
 */
public final class Orchestrator implements Runnable {

    private final PlayerQueue queue;
    private final ExecutorService pool;
    private final EventBus bus;
    private final Set<Match> activeMatches;
    private final long roundDelayMs;

    // Estatísticas
    private final long tournamentStartTime;
    private final AtomicLong totalMatchTimeNanos;
    private final AtomicInteger matchesPlayed;

    private int nextMatchId = 1;

    public Orchestrator(
            PlayerQueue queue,
            ExecutorService pool,
            EventBus bus,
            Set<Match> activeMatches,
            long roundDelayMs,
            long tournamentStartTime,
            AtomicLong totalMatchTimeNanos,
            AtomicInteger matchesPlayed
    ) {
        this.queue = queue;
        this.pool = pool;
        this.bus = bus;
        this.activeMatches = activeMatches;
        this.roundDelayMs = roundDelayMs;

        this.tournamentStartTime = tournamentStartTime;
        this.totalMatchTimeNanos = totalMatchTimeNanos;
        this.matchesPlayed = matchesPlayed;
    }

    @Override
    public void run() {
        try {

            while (true) {

                Player[] pair = queue.dequeuePair();

                if (pair == null) {
                    break;
                }

                Match match = new Match(
                        nextMatchId++,
                        pair[0],
                        pair[1]
                );

                activeMatches.add(match);

                bus.publish(
                        new Events.QueueChanged(
                                queue.snapshot(),
                                queue.aliveSnapshot()
                        )
                );

                pool.submit(
                        new MatchRunner(
                                match,
                                queue,
                                bus,
                                activeMatches,
                                roundDelayMs,
                                totalMatchTimeNanos,
                                matchesPlayed
                        )
                );
            }

        } catch (InterruptedException e) {

            Thread.currentThread().interrupt();

        } finally {

            pool.shutdown();

            // Tempo total real do torneio
            long totalTournamentTime =
                    System.nanoTime() - tournamentStartTime;

            bus.publish(
                    new Events.TournamentEnded(
                            queue.champion(),
                            totalTournamentTime,
                            totalMatchTimeNanos.get(),
                            matchesPlayed.get()
                    )
            );
        }
    }
}